lib/Frontend/ArgsToFrontendInputsConverter.cpp

```diff
@@ -30,33 +31,48 @@ using namespace swift;
 using namespace llvm::opt;
 
 ArgsToFrontendInputsConverter::ArgsToFrontendInputsConverter(
-    DiagnosticEngine &diags, const ArgList &args,
-    FrontendInputsAndOutputs &inputsAndOutputs)
-    : Diags(diags), Args(args), InputsAndOutputs(inputsAndOutputs),
+    DiagnosticEngine &diags, const ArgList &args)
+    : Diags(diags), Args(args),
       FilelistPathArg(args.getLastArg(options::OPT_filelist)),
       PrimaryFilelistPathArg(args.getLastArg(options::OPT_primary_filelist)) {}
 
-bool ArgsToFrontendInputsConverter::convert() {
+Optional<FrontendInputsAndOutputs> ArgsToFrontendInputsConverter::convert(
+    SmallVectorImpl<std::unique_ptr<llvm::MemoryBuffer>> *buffers) {
+  SWIFT_DEFER {
+    if (buffers) {
+      std::move(ConfigurationFileBuffers.begin(),
+                ConfigurationFileBuffers.end(),
+                std::back_inserter(*buffers));
+      // Clearing the original list of buffers isn't strictly necessary, but
+      // makes the behavior more sensible if we were to call convert() again.
+      ConfigurationFileBuffers.clear();
+    }
+  };
+
   if (enforceFilelistExclusion())
-    return true;
+    return None;
+
   if (FilelistPathArg ? readInputFilesFromFilelist()
                       : readInputFilesFromCommandLine())
-    return true;
+    return None;
   Optional<std::set<StringRef>> primaryFiles = readPrimaryFiles();
   if (!primaryFiles)
-    return true;
-  std::set<StringRef> unusedPrimaryFiles =
+    return None;
+
+  FrontendInputsAndOutputs result;
+  std::set<StringRef> unusedPrimaryFiles;
+  std::tie(result, unusedPrimaryFiles) =
       createInputFilesConsumingPrimaries(*primaryFiles);
 
   if (checkForMissingPrimaryFiles(unusedPrimaryFiles))
-    return true;
+    return None;
 
   // Must be set before iterating over inputs needing outputs.
-  InputsAndOutputs.setIsSingleThreadedWMO(isSingleThreadedWMO());
-
-  InputsAndOutputs.setBypassBatchModeChecks(
+  result.setIsSingleThreadedWMO(isSingleThreadedWMO(result));
```

Comment:
There's something funny here; the line reads oddly to my eye, as if it's checking the result and then setting the same thing it just checked for. `result` is being both checked and set, and `isSingleThreadedWMO` looks like it could be a static function here, too.
As I look at it, I think that this functionality might work better if moved to  the end of`createInputFilesConsumingPrimaries`, because that function creates and returns a `FrontendInputsAndOutputs`, so `createInputFilesConsumingPrimaries` can look at its state and do the computation.

Alternatively, this functionality could be moved into `FrontendInputsAndOutputs` as no-argument `cacheIsSingleThreadedWMO` function.

Reply:
The `createInputFilesConsumingPrimaries` worked!